src/meta/src/hummock/manager/mod.rs

```diff
@@ -878,243 +876,309 @@ impl HummockManager {
     }
 
     #[named]
-    pub async fn get_compact_task_impl(
+    pub async fn get_compact_tasks_impl(
         &self,
-        compaction_group_id: CompactionGroupId,
+        compaction_groups: Vec<CompactionGroupId>,
         selector: &mut Box<dyn CompactionSelector>,
-    ) -> Result<Option<CompactTask>> {
+    ) -> crate::hummock::error::Result<(Vec<CompactTask>, Vec<CompactTask>)> {
         // TODO: `get_all_table_options` will hold catalog_manager async lock, to avoid holding the
         // lock in compaction_guard, take out all table_options in advance there may be a
         // waste of resources here, need to add a more efficient filter in catalog_manager
+        let deterministic_mode = self.env.opts.compaction_deterministic_test;
         let all_table_id_to_option = self
             .metadata_manager
             .get_all_table_options()
             .await
             .map_err(|err| Error::MetaStore(err.into()))?;
-        let mut table_to_vnode_partition = match self
-            .group_to_table_vnode_partition
-            .read()
-            .get(&compaction_group_id)
-        {
-            Some(table_to_vnode_partition) => table_to_vnode_partition.clone(),
-            None => BTreeMap::default(),
-        };
 
         let mut compaction_guard = write_lock!(self, compaction).await;
         let compaction = compaction_guard.deref_mut();
-        let compaction_statuses = &mut compaction.compaction_statuses;
+        let mut versioning_guard = write_lock!(self, versioning).await;
```

Comment:
Prior to this PR, we don't acquire versioning write lock and only acquire a versioning https://github.com/risingwavelabs/risingwave/blob/e380a5d5651dc6ffa2a7f3f6dce7c29f0f026b32/src/meta/src/hummock/manager/mod.rs#L942 for `current_version.clone()`. I am concerned that holding versioning write lock here may cause contention between the batch `get_compact_task_impl` and `commit_epoch`. IIUC, the write lock here is only useful for applying changes for trivial move task and will not be needed if no trivial move task is generated. In this case, can we limit the scope of the write lock to only during applying trivial move tasks, not during the whole `get_compact_task_impl`? We can make the `gen_version_delta` logic incremental.

Reply:
This Pr is assume that most of calling `get_compact_task` will meet at lest once `trivial_task` (According our experience, the number of "trivial move task" is much larger than "normal task"). For behavior before this PR, every time "get_compact_task" also need to acquire write-lock for `versioning` if it gets a "trivial move task".